docs: add `README.md` to vendors/halo2 (#360)
kroma-network/tachyon: vendors/halo2/how_to_migrate_your_circuit.md

# How to migrate you circuit in Rust

Tachyon can replace the existing Halo2 implementations to generate proofs and currently supports *BN256 (or Bn254)* curve along with *GWC* and *SHPlonk* provers.

## Requirements

1. Replace `halo2` library in your program with [tachyon-halo2](https://github.com/kroma-network/halo2).

   ``` diff
   # In your Cargo.toml
   - halo2_proofs = { git = "https://github.com/zcash/halo2.git" }
   + halo2_proofs = { git = "https://github.com/kroma-network/halo2.git", branch = "halo2-with-tachyon" }
   ```

> 2. install the tachyon debian package, see *Debian packaging* section in [how_to_build.md](./how_to_build.md).

Reviewer: ```suggestion
2. Install the tachyon debian package. See *Debian packaging* section in [how_to_build.md](./how_to_build.md).
```

## Migration Steps

> 1. Creating Proving Key for Tachyon

Reviewer 2: ```suggestion
1. Create Proving Key for Tachyon
```

    ``` diff
      use halo2_proofs::plonk::halo2_proofs::{
    +     bn254::ProvingKey as TachyonProvingKey;
          halo2curves::bn256::{Bn256, Fr},
          plonk::keygen_pk2,
          poly::ParamsKZG,
      };

      // Use just your |circuit| and |instances| so.
      let circuit = YourCircuit<_>;
      let instances = <YourInstances>

      // |k|: log of degree, |s|: toxic-waste secret.
      let k: u32 = 5;
      let s = Fr::from(2);

      // Generate |params|.
      let params = ParamsKZG<Bn256>::unsafe_setup_with_s(k, s);
      let params: Option<ParamsKZG<Bn256>> = Some(params);

      // Generate |pk| as proving key.
      let pk = keygen_pk2(params.as_ref().unwrap(), &circuit).unwrap();
    + let mut tachyon_pk = {
    +     // Please remember to release |pk_bytes| immediately after
    +     // deserialization, as it can occupy big memory.
    +     let mut pk_bytes: Vec<u8> = vec![];
    +     pk.write(&mut pk_bytes, halo2_proofs::SerdeFormat::RawBytesUnchecked)
    +         .unwrap();
    +     TachyonProvingKey::from(pk_bytes.as_slice())
    + };
   ```

2. Change transcript for Tachyon

    ``` diff

>       use halo2_proofs::bn254::{
>     -     transcript::{Challenge255, PoseidonWrite},
>     +     PoseidonWrite as TachyonPoseidonWrite,
>       }

Reviewer 2: ```suggestion
    -  use halo2_proofs::transcript::{Challenge255, PoseidonWrite};
    -  use halo2curves::bn256::G1Affine;
    +  use halo2_proofs::bn254::PoseidonWrite as TachyonPoseidonWrite;
```

>     - let mut transcript = PoseidonWrite::<_, _, Challenge255<_>>::init(vec![]);

Reviewer 2: ```suggestion
    - let mut transcript = PoseidonWrite::<_, G1Affine, Challenge255<_>>::init(vec![]);
```

    + let mut transcript = TachyonPoseidonWrite::init(vec![]);
    ```

    Tachyon supports *PoseidonWrite*, *Blake2bWrite*, and *Sha256Write* as transcript methods.

3. Change RNG (Random Number Generator) for Tachyon

    ``` diff

>     - use rand_core::{RngCore, SeedableRng};
>     - use rand_xorshift::XorShiftRng;
>     + use halo2_proofs::xor_shift_rng::XORShiftRng;

Reviewer 2: ```suggestion
    - use rand_core::{RngCore, SeedableRng};
    - use rand_xorshift::XorShiftRng;
    + use halo2_proofs::xor_shift_rng::XORShiftRng;
    + use rand_core::SeedableRng;
```

      let seed: [u8; 16] = <SpecificSeed>;
    - let rng = rand_xorshift::XorShiftRng::from_seed(seed);
    + let rng = XORShiftRng::from_seed(seed);
    ```

4. Create Prover for Tachyon

    ``` diff
    + use halo2_proofs::{
    +     bn254::{
    +         GWCProver as TachyonGWCProver,
    +         TachyonProver,
    +     },
    +     consts::TranscriptType,
    +     halo2curves::bn256::Bn256,
    +     poly::kzg::commitment::KZGCommitmentScheme,
    + }
    +
    + let mut prover = {
    +     let mut params_bytes = vec![];
    +     // Please remember to release |params| immediately after
    +     // deserialization, as it can occupy big memory.
    +     params.unwrap().write(&mut params_bytes).unwrap();
    +     TachyonGWCProver::<KZGCommitmentScheme<Bn256>>::from_params(
    +         TranscriptType::Poseidon as u8,
    +         k,
    +         params_bytes.as_slice(),
    +     )
    + };
    ```

>     Tachyon supports GWCProver and SHPlonkProver as provers. And note that the first argument of `XXXProver::from_params()` and `transcript` must refer to the same type.

Reviewer 2: ```suggestion
    Tachyon supports `GWCProver` and `SHPlonkProver` as a prover. And note that the first argument of `XXXProver::from_params()` and `transcript` must refer to the same type.
```

> # How to migrate you circuit in Rust

Reviewer 2: ```suggestion
# How to migrate
```
